Dynamic Flexvolume plugin discovery, probing with filesystem watch.

verult opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: Enables dynamic Flexvolume plugin discovery. This model uses a filesystem watch (fsnotify library), which notifies the system that a probe is necessary only if something changes in the Flexvolume plugin directory.

This PR uses the dependency injection model in https://github.com/kubernetes/kubernetes/pull/49668.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #51470

**Release Note**:
```release-note
Dynamic Flexvolume plugin discovery. Flexvolume plugins can now be discovered on the fly rather than only at system initialization time.
```

/sig-storage

/assign @jsafrane @saad-ali 
/cc @bassam @chakri-nelluri @kokhang @liggitt @thockin

verult committed:
Adding dynamic Flexvolume plugin discovery capability, using filesystem watch.

liggitt commented on pkg/controller/volume/persistentvolume/pv_controller_base.go:
```diff
@@ -90,7 +90,7 @@ func NewController(p ControllerParameters) (*PersistentVolumeController, error)
 	}
 
-	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, controller); err != nil {
+	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, nil, controller); err != nil {
```
add `volume.DynamicPluginProber` to controllerparameters?

verult replied:
AFAIK PV controller was never aware of Flexvolume.

verult commented:
Added filesystem watch implementation of the prober.

kokhang commented:
@verult I tested this locally and it did work. Kubelet was able to load the Rook flexvolume without having to restart. Here is the log:
`I0808 02:22:21.564582   17759 plugins.go:420] Loaded volume plugin "flexvolume-kubernetes.io/rook"`

However there are still some concerns that i have listed in https://github.com/kubernetes/community/pull/833/. Namely about how to obtain or construct a client config so that the driver can talk to the K8S api.

verult commented:
/cc @msau42

saad-ali commented on pkg/volume/flexvolume/probe.go:
```diff
@@ -19,28 +19,185 @@ package flexvolume
+		}
+	}
+
+	if time.Since(prober.lastUpdated) > time.Second { // Reduce the frequency of probes.
```
nit: add to const at top of file. Maybe make it less frequent... 10 seconds? Want to avoid frequent FS reads.

saad-ali commented on pkg/volume/flexvolume/probe.go:
```diff
@@ -19,28 +19,185 @@ package flexvolume
+	}
+
+	if eventOpIs(event, fsnotify.Remove) && eventPathAbs == pluginDirAbs {
+		// pluginDir needs to exist in order to be watched.
```
Post a warning to log.

saad-ali commented on pkg/volume/flexvolume/probe.go:
```diff
@@ -19,28 +19,185 @@ package flexvolume
+		return err
+	}
+
+	if eventOpIs(event, fsnotify.Remove) && eventPathAbs == pluginDirAbs {
```
Add a comment above this block indicating what you're doing and why.

saad-ali commented on pkg/volume/flexvolume/probe.go:
```diff
@@ -19,28 +19,185 @@ package flexvolume
+		prober.lastUpdated = time.Now()
+	}
+
+	return nil
```
There is a race here if files are dropped quickly in succession, and the read happens before the 2nd event is processed.

Let's make sure to add testing to see how often this might happen, and if it is infrequent put a comment indicating you know about it.

verult replied:
Fixing this by adding a limit on number of events processed per second, and ignoring events from .dot files.

Users need to install drivers atomically, so if there are ever multiple files in an installation, O(1) events are triggered regardless. Multiple events are triggered only when multiple Flexvolume drivers are installed.

saad-ali reviewed (changes requested):
Took a first pass

kokhang commented:
I know this is unrelated but what do you recommend to deploy the flexvolume drivers on nodes that are not configured to have the same `flexvolumeDriverDir`? I followed the proposed example using daemon set. But i was thinking that this would not work in the case where the driverDir is different in  other nodes.

verult commented:
@kokhang There was some discussion about this topic here: https://github.com/kubernetes/community/pull/833#discussion_r130722002

verult commented:
Addressed comments and refactored test files in separate commits.

saad-ali reviewed (commented):
We spoke offline: See if it is possible to support atomic installation of a driver with additional files by only registering the driver when the driver file shows up (instead of just the driver directory).

verult commented:
Addressed comments.

As discussed offline, I confirmed that driver can only be recognized when file matches the expected driver name based on the directory name.

verult commented:
/retest pull-kubernetes-e2e-gce-etcd3

verult commented:
/test pull-kubernetes-e2e-gce-etcd3

verult commented:
/test pull-kubernetes-e2e-kops-aws

saad-ali commented:
> Addressed comments.

Did you push the latest changes?

verult commented:
Some stuff got lost in all the refactoring... but they are back in now.

verult commented:
/release-note

kokhang commented:
@verult @saad-ali I have encountered an issue with the flexvolume approach and I am wondering  if you have some guidance on how to proceed with this. 

I have deployed K8S using Kubeadm. Kubeadm deploys the api, scheduler and controller-manager in a container. The issue with this is that when flexvolume is using the attacherdetacher-controller, the controller-manager manifest needs to have the flexvolume dir mounted from the hostpath. So even restarting kubelet would not solve this case.

If you add the hostpath mount and restart the controller-manager pod then it works. But it is an invasive procedure.

liggitt commented:
I would recommend reporting the issue to kubeadm as a bug… I wouldn't expect that to be a desired deployment configuration

kokhang commented:
@liggitt ill contact to the kubeadm folks. But im also wondering if the controller-manager needs to probe for the flexvolume plugin as well? Or is this case already handled by kubelet probe?

verult commented:
In order to perform attach-detach operations, the controller-manager would have to probe for Flexvolume driver. One possible solution for now, while waiting for a response from the kubeadm team, is to use one of controller-manager pod's existing hostpath. I'm not sure if the kubeadm installation uses this but, /etc/srv/kubernetes might be a possibility.

kokhang commented:
> In order to perform attach-detach operations, the controller-manager would have to probe for Flexvolume driver.

Does this mean that this dynamic flexvolume discovery fix is only applicable for non attacher-detacher-controller based flexvolume plugins?

kokhang commented:
I just saw the code. It seems you are doing the probe in the controller-manager as well

saad-ali commented:
/lgtm
/approve

saad-ali commented:
/lgtm
